lib/index.ts

```diff
@@ -87,19 +87,35 @@ export async function fetchBeaconByTime(client: ChainClient, time: number): Prom
 }
 
 // an async generator emitting beacons from the latest round onwards
-export async function* watch(client: ChainClient, abortController: AbortController): AsyncGenerator<RandomnessBeacon> {
-    while (!abortController.signal.aborted) {
-        const info = await client.chain().info()
-        const beacon = await client.latest()
-        yield validatedBeacon(client, beacon)
+export async function* watch(
+    client: ChainClient,
+    abortController: AbortController,
+    options: WatchOptions = defaultWatchOptions,
+): AsyncGenerator<RandomnessBeacon> {
+    const info = await client.chain().info()
+    let currentRound = roundAt(Date.now(), info)
 
-        const nextRoundTime = roundTime(info, beacon.round + 1)
+    while (!abortController.signal.aborted) {
         const now = Date.now()
+        await sleep(roundTime(info, currentRound) - now)
 
-        await sleep(nextRoundTime - now)
+        const beacon = await retryOnError(async () => client.get(currentRound), options.retriesOnFailure)
+        yield validatedBeacon(client, beacon)
+        currentRound = currentRound + 1
     }
 }
 
+// options for the async generator `watch` function
+export type WatchOptions = {
+    // the number of retries to attempt for a failed response
+    // could be useful if e.g. the network does not generate a round perfectly on time
+    retriesOnFailure: number
+}
+
+const defaultWatchOptions = {
+    retriesOnFailure: 3
+}
```

Comment:
Isn't 3 a pretty low value here? I mean if the server responds 404 quickly, the 3 attempts are reached easily, especially since `retryOnError` has no sleep between the tries. E.g.
- 404 after 150ms
- 404 after 300ms
- 404 after 450ms
- 404 after 600ms
should trigger a watch exception.

In my testing this does not happen currently. I get beacons 800-1000ms after publishing time. But I guess this just works because the error responses are slow.

Reply:
Okay, the reason this works is the HTTP endpoint allows long-polling for the next round. The first request succeeds even if it had to wait multiple seconds. This can be reproduces in command line like this:

```
$ time curl -v "https://api.drand.sh/8990e7a9aaed2ffed73dbd7092123d6f289930540d7651336225dc172e51b2ce/public/2626798"
{"round":2626798,"randomness":"de4bfb39631eb6e4d03af173c231c049725808df69872f3b76d262d597aec8ab","signature":"b0e48b873dc8a5ce90419da58b4d4cdd466d74609a4ee65fb3bb1bcbcb2fc2ba200c55a60b24966546d4979e0db9aaa30e625ef8782b91cb6f9824cf2de713a46b0ce1d2fed526551363ec18e0453e65e9b0e966fe505bef3e1f72bd00227bc6","previous_signature":"87cd3effc8b99462310e95e6cdc0ae3c1dc3470587e4bc7b6f5ff643217aa0b881ea3ac6705995649b4f35049306101d194567617ac768cecbc36b74d99b4ed03eb2932fbabf4523779ba5d76556b68940ccee3b78154f722ca51b3225b8420e"}curl -v   0,02s user 0,01s system 0% cpu 13,725 total
```

However, I'm not sure if this is something we can rely on 🤷‍♂️

Reply:
Yeah, that's also something I brought up in an above comment, but  we're honoring requests for the nest round on the server side.